pkg/apis/kudo/v1beta1/instance_admission.go

```diff
@@ -0,0 +1,239 @@
+	switch {
+	case hasPlan && isParameterUpdate && *triggeredPlan != oldPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: plan '%s' is scheduled and an update would trigger a different plan '%s'", old.Namespace, old.Name, oldPlan, newPlan)
+	case isUpgrade && hasPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s while a plan '%s' is scheduled is not allowed", old.Namespace, old.Name, newOvRef, oldPlan)
+	case isUpgrade && isNovelPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s and triggering new plan '%s' is not allowed", old.Namespace, old.Name, newOvRef, newPlan)
+	case isPlanOverride:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: overriding currently scheduled plan '%s' with '%s' is not supported", old.Namespace, old.Name, oldPlan, newPlan)
+	case isPlanCancellation:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: cancelling currently scheduled plan '%s' is not supported", old.Namespace, old.Name, oldPlan)
+	case isParameterUpdate && isUpgrade:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: upgrade to new OperatorVersion %s together with a parameter update triggering '%s' is not allowed", old.Namespace, old.Name, newOvRef, *triggeredPlan)
+	case isParameterUpdate && isNovelPlan:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: triggering one plan '%s' directly and through parameter update '%s' is not allowed", old.Namespace, old.Name, oldPlan, newPlan)
+	case isParameterUpdate && isPlanOverride:
+		return nil, fmt.Errorf("failed to update Instance %s/%s: updating parameters and triggering plan '%s' is not allowed", old.Namespace, old.Name, *triggeredPlan)
+	}
+
+	// Deciding which plan to trigger:
+	switch {
+	case isUpgrade:
+		plan := SelectPlan([]string{UpgradePlanName, UpdatePlanName, DeployPlanName}, ov)
+		if plan == nil {
+			return nil, fmt.Errorf("failed to update Instance %s/%s: couldn't find any suitable plan that would be triggered by an OperatorVersion upgrade", old.Namespace, old.Name)
+		}
+		return plan, nil
+
+	case isParameterUpdate:
+		// if the same plan is triggered by the update, we clean the Instance.Status to effectively restart the plan
+		if *triggeredPlan == oldPlan {
+			if err := resetInstanceStatus(old); err != nil {
+				return nil, fmt.Errorf("failed to update Instance %s/%s: %v", old.Namespace, old.Name, err)
+			}
+		}
+		return triggeredPlan, nil
+
+	case isNovelPlan:
+		return &newPlan, nil
+		// Implement plan overrides and cancellations below:
+	}
+
+	return triggeredPlan, nil
+}
+
+// resetInstanceStatus clears Instance.Status to effectively restart existing plan
+func resetInstanceStatus(instance *Instance) error {
+	// TODO (AD): implement
+	return nil
+}
+
+// getOperatorVersion retrieves operator version belonging to the given instance
+func (ia *InstanceAdmission) getOperatorVersion(instance *Instance) (ov *OperatorVersion, err error) {
```

Comment:
I almost feel like this method can be in kudo client as many code paths need it

Reply:
Good catch. We copy-pasted it from the [instance_controller](https://github.com/kudobuilder/kudo/blob/master/pkg/controller/instance/instance_controller.go#L357) back then because there it is a method of the `Reconciler`. I'll refactor it.